<a href="https://colab.research.google.com/github/CarmenTheodoraCraciun/MNIST-Handwritten-digits/blob/main/SIIA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prepararea datelor

In [1]:
import pandas as pd
df_test = pd.read_csv("/content/sample_data/mnist_test.csv", header=None)
df_test.head()

0    1    2    3    4    5    6    7    8    9    ...  775  776  777  778  \
0    7    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
1    2    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
2    1    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   
4    4    0    0    0    0    0    0    0    0    0  ...    0    0    0    0   

   779  780  781  782  783  784  
0    0    0    0    0    0    0  
1    0    0    0    0    0    0  
2    0    0    0    0    0    0  
3    0    0    0    0    0    0  
4    0    0    0    0    0    0  

[5 rows x 785 columns]

In [2]:
# Definitia antetelor daca acestea nu exista
column_names = ["label"] + [f"{i}x{j}" for i in range(1, 29) for j in range(1, 29)]
df_test.columns = column_names
df_test.head()

label  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  28x19  28x20  \
0      7    0    0    0    0    0    0    0    0    0  ...      0      0   
1      2    0    0    0    0    0    0    0    0    0  ...      0      0   
2      1    0    0    0    0    0    0    0    0    0  ...      0      0   
3      0    0    0    0    0    0    0    0    0    0  ...      0      0   
4      4    0    0    0    0    0    0    0    0    0  ...      0      0   

   28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
0      0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0      0  

[5 rows x 785 columns]

In [5]:
df_train = pd.read_csv("/content/sample_data/mnist_train.csv")
column_names = ["label"] + [f"{i}x{j}" for i in range(1, 29) for j in range(1, 29)]
df_train.columns = column_names
df_train.head()

label  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  28x19  28x20  \
0      5    0    0    0    0    0    0    0    0    0  ...      0      0   
1      7    0    0    0    0    0    0    0    0    0  ...      0      0   
2      9    0    0    0    0    0    0    0    0    0  ...      0      0   
3      5    0    0    0    0    0    0    0    0    0  ...      0      0   
4      2    0    0    0    0    0    0    0    0    0  ...      0      0   

   28x21  28x22  28x23  28x24  28x25  28x26  28x27  28x28  
0      0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0      0  

[5 rows x 785 columns]

In [6]:
from sklearn import svm, metrics
from sklearn.model_selection import train_test_split

# Separam etichetele de pixeli
X_train = df_train.drop('label', axis=1)
y_train = df_train['label']
print(X_train)
print(y_train)

       1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  1x10  ...  28x19  28x20  \
0        0    0    0    0    0    0    0    0    0     0  ...      0      0   
1        0    0    0    0    0    0    0    0    0     0  ...      0      0   
2        0    0    0    0    0    0    0    0    0     0  ...      0      0   
3        0    0    0    0    0    0    0    0    0     0  ...      0      0   
4        0    0    0    0    0    0    0    0    0     0  ...      0      0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  ...    ...    ...   
19994    0    0    0    0    0    0    0    0    0     0  ...      0      0   
19995    0    0    0    0    0    0    0    0    0     0  ...      0      0   
19996    0    0    0    0    0    0    0    0    0     0  ...      0      0   
19997    0    0    0    0    0    0    0    0    0     0  ...      0      0   
19998    0    0    0    0    0    0    0    0    0     0  ...      0      0   

       28x21  28x22  28x23  28x24  28x25  28x26  28

In [7]:
X_test = df_test.drop('label', axis=1)
y_test = df_test['label']

# Functia de evaluare a modelelor

In [8]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

def evaluate_model(y_true, y_pred):
    '''
    Funcția este concepută pentru a evalua performanța unui model de clasificare prin calcularea diferitelor valori, cum ar fi acuratețea, raportul de clasificare și matricea de confuzie.

    Parametri:
       - y_true = Etichetele adevărate sau valorile de adevăr de bază.
       - y_pred = Etichetele estimate generate de model.
    '''
    # Calculeaza acuratetea
    accuracy = accuracy_score(y_true, y_pred)
    print("Accuracy:", accuracy)

    # Raportul de clasificare
    print("Classification report:")
    print(classification_report(y_true, y_pred))

    # Matricea de confuzie
    print("Confusion matrix:")
    print(confusion_matrix(y_true, y_pred))

* **Acuratețea** este raportul dintre numărul de predicții corecte și numărul total de predicții.
* **Raportul de clasificare** oferă o analiză detaliată a performanței modelului pentru fiecare clasă. Include metrici precum precizia, acoperirea (recall) și f1-score pentru fiecare clasă, precum și suportul (numărul de instanțe din fiecare clasă).
  * *Precizia* - rata de predicții corecte pentru o anumită clasă.
  $precision = \frac{True Positives}{True Positives + False Positives}$
  * *Acoperire* - rata de identificare a instanțelor relevante pentru o anumită clasă.
  $recall = \frac{True Positives}{True Positives + False Negative}$
  * *F1-score* - o măsură a performanței unui model de clasificare și este definită ca media armonică între precizie și acoperire (recall). Această măsură este utilă în special atunci când există un dezechilibru între clase în setul de date.
  $f1Score=2\frac{precision * recall}{precision + recall}$
  * *Numarul de instanțe* din setul de date de testare care aparțin unei clase.
  * *Macro Average* - media aritmetică a preciziilor, acoperirii și F1-score-urile calculate pentru fiecare clasă individuală. Aceasta nu ține cont de dezechilibrul dintre clase și tratează fiecare clasă în mod egal în calculul mediei.
  * *Weighted Average* - similar cu media macro, însă fiecare clasă este ponderată în funcție de numărul de exemple din acea clasă în setul de date de testare. Acest lucru înseamnă că clasele cu un număr mai mare de exemple vor avea o influență mai mare asupra valorii medii.
* **Matricea de confuzie** este o matrice în care fiecare rând reprezintă clasa reală, iar fiecare coloană reprezintă clasa prezisă. Elementul (i, j) al matricei este numărul de instanțe ale clasei i care au fost clasificate ca aparținând clasei j. Diagonala principală conține numărul de clasificări corecte pentru fiecare clasă.

# Random Forest

Random Forest este un algoritm de învățare supervizată care combină mai mulți arbori de decizie pentru a obține predicții mai precise. Folosește o tehnică numită "bagging" și construiește arbori de decizie pe subseturi aleatorii ale setului de date. Acest lucru îl face robust și rezistent la overfitting. De asemenea, poate estima importanța fiecărei caracteristici din setul de date.

In [9]:
from sklearn.ensemble import RandomForestClassifier

# Cream clasificatorul
classifier = RandomForestClassifier(n_estimators=100)

# Antrenam modelul
classifier.fit(X_train, y_train)

# Efectuarea de predicții asupra datelor de testare
y_pred = classifier.predict(X_test)

evaluate_model(y_test, y_pred)

Accuracy: 0.9589
Classification report:
              precision    recall  f1-score   support

           0       0.96      0.99      0.97       980
           1       0.99      0.99      0.99      1135
           2       0.94      0.96      0.95      1032
           3       0.95      0.94      0.95      1010
           4       0.96      0.96      0.96       982
           5       0.97      0.95      0.96       892
           6       0.96      0.97      0.97       958
           7       0.96      0.94      0.95      1028
           8       0.95      0.94      0.95       974
           9       0.94      0.94      0.94      1009

    accuracy                           0.96     10000
   macro avg       0.96      0.96      0.96     10000
weighted avg       0.96      0.96      0.96     10000

Confusion matrix:
[[ 968    0    1    0    0    3    3    1    4    0]
 [   0 1121    4    3    0    1    3    1    2    0]
 [   6    0  987    8    5    0    5   13    8    0]
 [   1    0   18  954   

Modelul a prezis corect cifra din imaginile de testare în aproximativ 95.85% din cazuri.

Din raportul de clasificare, pentru cifra:
* 0 avem o precizie de 97%, o acoperire de 99%, iar media armonică dintre precizie și acoperire este de 98% din 980 de instanțe.
* 5 avem precizie de 97%, o acoperire de 95%, iar media armonică dintre precizie și acoperire este de 96% din 892 de instanțe.

Acuratețea modelul este una bună, dar vom încerca un alt model pentu una și mai bună.

# DNN - Dense Neural Networks

Rețelele neuronale dense sunt o clasă de rețele neuronale artificiale în care neuroni (unități de procesare) sunt aranjați în straturi. Fiecare neuron dintr-un strat este conectat la toți neuroni din stratul anterior și la toți neuroni din stratul următor. Aceste conexiuni "dense" între straturi sunt ceea ce definește rețelele neuronale dense.

DNN-urile sunt capabile să învețe caracteristici complexe din date, făcându-le utile pentru o varietate de sarcini, cum ar fi recunoașterea imaginilor, clasificarea textului, regresia etc.

Deși DNN-urile sunt puternice, ele pot fi predispuse la overfitting (modelul învață să identifice modele irelevante din datele de antrenament, ceea ce duce la o performanță slabă pe datele noi) pe seturile de date mici sau zgomotoase.

In [13]:
!pip install tensorflow

In [14]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Flatten
from tensorflow.keras.layers import Conv2D, MaxPooling2D

# Convertim cadrele de date in matrice
# Redimensionam datele pentru a se potrivi cu formularul de introducere a modelului (num_samples, 28, 28, 1)
X_train = X_train.values.reshape(-1, 28, 28, 1)
X_test = X_test.values.reshape(-1, 28, 28, 1)

# Normalizam datele
X_train, X_test = X_train / 255.0, X_test / 255.0

In [15]:
# Cream modelul
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

In [16]:
# Compilam
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Antrenam
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
625/625 [==============================] - 71s 106ms/step - loss: 0.3000 - accuracy: 0.9074
Epoch 2/5
625/625 [==============================] - 57s 91ms/step - loss: 0.1135 - accuracy: 0.9646
Epoch 3/5
625/625 [==============================] - 56s 90ms/step - loss: 0.0785 - accuracy: 0.9760
Epoch 4/5
625/625 [==============================] - 59s 95ms/step - loss: 0.0676 - accuracy: 0.9785
Epoch 5/5
625/625 [==============================] - 56s 90ms/step - loss: 0.0519 - accuracy: 0.9841


In [17]:
y_pred = model.predict(X_test)
y_pred = np.argmax(y_pred, axis=1)

313/313 [==============================] - 6s 20ms/step


In [18]:
evaluate_model(y_test, y_pred)

Accuracy: 0.9847
Classification report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98       980
           1       0.99      1.00      0.99      1135
           2       0.99      0.99      0.99      1032
           3       0.99      0.98      0.99      1010
           4       0.99      0.98      0.99       982
           5       1.00      0.98      0.99       892
           6       0.98      0.98      0.98       958
           7       0.97      0.99      0.98      1028
           8       0.99      0.96      0.98       974
           9       0.99      0.98      0.98      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000

Confusion matrix:
[[ 976    0    0    0    0    0    1    1    1    1]
 [   0 1134    0    0    0    0    1    0    0    0]
 [   1    3 1021    0    0    0    0    6    1    0]
 [   0    0    4  994   

Modelul a prezis corect cifra din imaginile de testare în aproximativ 98.47% din cazuri.

Din raportul de clasificare, pentru cifra:
* 0 avem o precizie de 97%, o acoperire perfectă, iar media armonică dintre precizie și acoperire este de 98% din 980 de instanțe.
* 5 avem precizie perfectă, o acoperire de 98%, iar media armonică dintre precizie și acoperire este de 99% din 892 de instanțe.

Din matricea de confuzie, pentru cifra:
* 0 înțelegem că modelul a recunoscut:
  * 976 de predicții corecte, adică cifra 0 a fost de 976 corect clasificată (cazuri adevărat pozitive);
  * câte o predicție greșită ca 6, 7, 8 și 9, adică cifra 0 a fost intuită în câte un caz ca cele patru cifre enumerate (cazuri fals negative);
  * greșit câte o singură dată cifrele 2, 4 și 7 ca fiind cifra 0(cazuri fals pozitive);
  * de două ori a interpretat greșit cifra 5 ca cifra 0 (cazuri fals pozitive);
  * de patru ori a intui cifra 4 ca cifra 0 (cazuri fals pozitive);
  * a intuit cifrele 6 și 8 ca 0 de 12 și respectiv 14 ori (cazuri fals pozitive).
  

In [29]:
y_pred

array([7, 2, 1, ..., 4, 5, 6])

In [30]:
import pandas as pd

# Convertim y_pred intr-un array de string-uri
y_pred_str = [str(i) for i in y_pred]

# Convertim y_pred_str intr-un DataFrame
df = pd.DataFrame({'Label': y_pred_str})

# Adaugam un index care incepe de la 1
df.index = range(1, len(df) + 1)

# Adaugam indexul ca o noua coloana 'ImageId'
df.reset_index(level=0, inplace=True)
df = df.rename(columns = {'index':'ImageId'})

# Exportam DataFrame-ul intr-un fisier CSV
df.to_csv('trupa_pederda_pred.csv', index=False)